tests/ModelTest.php

```diff
@@ -325,11 +329,106 @@ public function testSoftDelete(): void
         $this->assertEquals(2, Soft::count());
     }
 
-    public function testPrimaryKey(): void
+    /**
+     * @dataProvider provideId
+     */
+    public function testPrimaryKey(string $model, $id, string $type): void
     {
-        $user = new User;
+        $model::truncate();
+
+        $document = new $model;
+        $this->assertEquals('_id', $document->getKeyName());
+
+        $document->_id = $id;
+        $document->save();
+        $this->assertSame($type, get_debug_type($document->_id));
+        $this->assertEquals($id, $document->_id);
+        $this->assertSame($type, get_debug_type($document->getKey()));
+        $this->assertEquals($id, $document->getKey());
+
+        $check = $model::find($id);
+
+        $this->assertNotNull($check, 'Not found');
+        $this->assertSame($type, get_debug_type($check->_id));
+        $this->assertEquals($id, $check->_id);
+        $this->assertSame($type, get_debug_type($check->getKey()));
+        $this->assertEquals($id, $check->getKey());
+    }
+
+    public static function provideId(): iterable
+    {
+        yield 'int' => [
```

Comment:
I think it would also make sense to check the expected value of the identifier, not only the type. This is relevant as the default behaviour for binary UUIDs is to cast it to a string, which is completely different from what the cast does.

Reply:
Updated the tests to validate the expected value. Using lose `assertEquals` because the `ObjectId` instance is recreated when the "model" is hydrated from the database.